ocaml/xenopsd/xc/service.ml

```diff
@@ -256,31 +265,66 @@ module DaemonMgmt (D : DAEMONPIDPATH) = struct
-          Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
-              try
-                Unix.lockf fd Unix.F_TRLOCK 0 ;
-                (* we succeeded taking the lock: original process is dead.
-                 * some other process might've reused its pid *)
-                None
-              with Unix.Unix_error (Unix.EAGAIN, _, _) ->
-                (* cannot obtain lock: process is alive *)
-                Some pid
-          )
-      | Xenstore key | Path_of {key; _} ->
-          (* backward compatibility during update installation: only has
-             xenstore pid available *)
-          let pid = xs.Xs.read (key domid) in
-          Some (int_of_string pid)
+      let cmdline_str =
+        Printf.sprintf "/proc/%d/cmdline" pid |> Unixext.string_of_file
+      in
+      match cmdline_str with
+      | "" ->
+          (* from man proc:
+             /proc/[pid]/cmdline
+             This read-only file holds the complete command line for the process,
+             unless the process is a zombie. In the latter case, there is nothing
+             in this file: that is, a read on this file will return 0 characters.
+             This applies when the VM is being shut down. *)
+          false
       | _ ->
-          None
-    with _ -> None
+          let cmdline = Astring.String.cuts ~sep:"\000" cmdline_str in
+          let valid =
+            List.for_all (fun arg -> List.mem arg cmdline) expected_args
+          in
+          if not valid then
+            error "%s: pid read from %s not valid (pid = %d)" D.name pid_source
+              pid ;
+          valid
+    with _ -> false
+
+  let pid ~xs domid =
+    let ( let* ) = Option.bind in
+    let* pid, pid_source =
+      try
+        match D.pid_location with
+        | (File file | Path_of {file; _}) when Sys.file_exists (file domid) ->
+            let path = file domid in
+            let* pid = Unixext.pidfile_read path in
+            Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
+                try
+                  Unix.lockf fd Unix.F_TRLOCK 0 ;
+                  (* we succeeded taking the lock: original process is dead.
+                   * some other process might've reused its pid *)
+                  None
+                with Unix.Unix_error (Unix.EAGAIN, _, _) ->
+                  (* cannot obtain lock: process is alive *)
+                  Some (pid, "pidfile")
+            )
+        | Xenstore key | Path_of {key; _} ->
+            (* case 1: backward compatibility during update installation: only has
+                 xenstore pid available.
+               case 2: pidfile(file domid) got removed unexpectly *)
+            let pid = xs.Xs.read (key domid) in
+            Some (int_of_string pid, "xenstore")
+        | _ ->
+            None
+      with _ -> None
+    in
+    if is_cmdline_valid ~pid ~pid_source (D.expected_cmdline_items ~domid) then
+      Some pid
+    else
+      None
```

Comment:
Should this not be logged? We failed to find the PID and but there is no explanations in the log. I think we should have reasons why we could not find a pid. That requires logging earlier - not just here.

Reply:
Oh, sorry, see it just now! I think you are right, to make future debug easy, we need log inside `is_cmdline_valid` whenever `false` is returned.